File: algorithms/1114.print-in-order/sotlution.py
from threading import Thread, Lock

class Foo(object):
    def __init__(self):
        self.lock_second = Lock()
        self.lock_third = Lock()

        self.lock_second.acquire()
        self.lock_third.acquire()


    def first(self, printFirst):
        """
        :type printFirst: method
        :rtype: void
        """
        
        # printFirst() outputs "first". Do not change or remove this line.
        printFirst()
        self.lock_second.release()


    def second(self, printSecond):
        """
        :type printSecond: method
        :rtype: void
        """
        self.lock_second.acquire()
        # printSecond() outputs "second". Do not change or remove this line.
        printSecond()
        self.lock_third.release()


    def third(self, printThird):
        """
        :type printThird: method
        :rtype: void
        """
        self.lock_third.acquire()
        # printThird() outputs "third". Do not change or remove this line.
        printThird()

File: algorithms/1114.print-in-order/sotlution.cpp
#include <mutex>

using namespace std;

class Foo {
    mutex mtx_second;
    mutex mtx_third;

  public:
    Foo() {
        mtx_second.lock();
        mtx_third.lock();
    }

    void first(function<void()> printFirst) {

        // printFirst() outputs "first". Do not change or remove this line.
        printFirst();
        mtx_second.unlock();
    }

    void second(function<void()> printSecond) {
        mtx_second.lock();
        // printSecond() outputs "second". Do not change or remove this line.
        printSecond();
        mtx_third.unlock();
    }

    void third(function<void()> printThird) {
        mtx_third.lock();
        // printThird() outputs "third". Do not change or remove this line.
        printThird();
    }
};
